Proposal to separate context propagation from observability (#42)
open-telemetry/oteps: text/0000-separate-context-propagation.md

# Proposal: Separate Layer for Context Propagation 

Status: `proposed`

Design OpenTelemetry as a set of separate applications which operate on a shared context propagation mechanism.

## Motivation

Based on prior art, we know that fusing the observability system and the context propagation system together creates issues. Observability systems have special rules for propagating information, such as sampling, and may have different requirements from other systems which require non-local information to be sent downstream. 
* Separation of concerns
  *   Remove the Tracer dependency from context propagation mechanisms.
  *   Separate distributed context into Baggage and Correlations
* Extensibility
  *   Allow users to create new applications for context propagation.
  *   For example: A/B testing, encrypted or authenticated data, and new, experimental forms of observability.

## Explanation

# OpenTelemetry Layered Architecture

> ![drawing](img/context_propagation_explanation.png)

Reviewer: I find this pic very confusing, sorry. There was an ascii art in one of the earlier tickets.

OpenTelemetry is a distributed program, which requires non-local, transaction-level context in order to execute correctly. Transaction-level context can also be used to build other distributed programs, such as security, versioning, and network switching programs.

To allow for this extensibility, OpenTelemetry is separated into **application layer** and a **context propagation layer**. In this architecture, multiple distributed applications - such as the observability and baggage systems provided by OpenTelemetry - simultaneously share the same underlying context propagation system in order to execute their programs.

# Application Layer

## Observability API

OpenTelemetry currently contains two observability systems - Tracing and Metrics – and may be extended over time. These separate systems are bound into a unified Observability API through sharing labels – a mechanism for correlating independent observations – and through sharing propagators.

**Observe(context, labels…, observations...) -> context**  

> The general form for all observability APIs is a function which takes a Context, label keys, and observations as input, and returns an updated Context.

Reviewer 2: what is an observation in this context? A single metric measurement comes to mind.

Reviewer 3: This describes

> The general form for all observability APIs

In my understanding, it is a stand-in for e.g. starting, ending Spans, but also any metric recording.

Reviewer 2: I guess it's still not something that immediately clicks for me. Some clarification on that, if it's not around in some other document, would be great.

**Correlate(context, label, value, hoplimit) -> context**  
To set the label values used by all observations in the current transaction, the Observability API provides a function which takes a context, a label key, a value, and a hoplimit, and returns an updated context. If the hoplimit is set to NO_PROPAGATION, the label will only be available to observability functions in the same process. If the hoplimit is set to UNLIMITED_PROPAGATION, it will be available to all downstream services.

> **GetPropagator(type) -> (inject, extract)**  

Reviewer: The `GetPropagator` propagator signature is confusing. The main idea of extensible propagation is that the instrumentation code that performs injection/extraction should not need to know/care about which types of contexts are being serialized. The only way this could work is if there is a shared "propagator" object that application interacts with, and different context sub-systems register their own serializers with that propagator. There's no situation where something would be calling GetPropagator on individual sub-systems like Correlations / Baggage.

Reviewer 3: Maybe the implementations of Inject and Extract would need it?

To register with the propagation system, the Observability API provides a set of propagation functions for every propagation type. 

## Baggage API

In addition to observability, OpenTelemetry provides a simple mechanism for propagating arbitrary data, called Baggage. This allows new distributed applications to be implemented without having to create new propagators.

To manage the state of a distributed application, the Baggage API provides a set of functions which read, write, and remove data.

> **SetBaggage(context, key, value) -> context**  

Reviewer 2: for these APIs, is it just returning back the instance that was passed in? or is it actually creating a copy? 

I'm not sure if it would be efficient for a language like python to rapidly create a bunch of new, but similar, objects.

Reviewer 3: I think this was just meant as a sort of mathematical notation, and whether this is implemented as a function modifying a thread-local it does not even get as argument or having a signature that actually has a context as parameter and return value is a language-specific concern.

> **SetBaggage(context, key, value) -> context**  

Reviewer: How do we want to treat types of the arguments here? Specifically, `value` would be more efficiently represented as `[]byte` rather than a string (recall J.Mace's baggage as thrift-like structs).

> To record the distributed state of an application, the Baggage API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

Reviewer 4: Would it also be beneficial to have a `hoplimit` on baggage items? For example, once supported, we may want a certain baggage value only to be propagated to the downstream service and no further - or only available within the service (`NO_PROPAGATION`) - but it is not to be treated as a label.

Reviewer: I've been in general skeptical about the TTL notion, especially "single-hop" whose semantics are completely vague given that requests may pass through proxies that the caller would not know about.

There is always the underlying in-memory Context object used as a storage for all kinds of contexts. I can always add a value to that context _explicitly_, and retrieve it _explicitly_ if my application needs it in-process. I don't think there's a need to provide any additional functionality here (although it's worth calling out that such capability must exist in the Context, whereas it was missing from OpenTracing Java, for example).

Author: Yeah, I agree that single hop TTL is questionable...

@objectiser I agree that if we come up with a TTL scheme beyond just `NO_PROPAGATION` and `UNLIMITED_PROPAGATION` we should add it to baggage. I only left it off because `NO_PROPAGATION` is equivalent to just using the context API directly, and that did not leave any options other than `UNLIMITED_PROPAGATION`. So it wasn't clear to me why anyone would need baggage for that purpose.

**GetBaggage(context, key) -> value**  
To access the distributed state of an application, the Baggage API provides a function which takes a context and a key as input, and returns a value.

**RemoveBaggage(context, key) -> context**  
To delete distributed state from an application, the Baggage API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

**ClearBaggage(context) -> context**  
To avoid sending baggage to an untrusted downstream process, the Baggage API provides a function remove all baggage from a context, 

> **GetPropagator(type) -> (inject, extract)**  

Reviewer 3: Which "different interfaces" do you mean? Did you mean to write "different interfaces for different propagators"?

> **GetPropagator(type) -> (inject, extract)**  

Reviewer 2: Sorry, interfaces isn't a great description. I was basically saying the same service may need to contain multiple different propagation configurations. Expanding on my example above, I could have two http clients: one for internal services (which should propagate headers) and one for external services (which should not, and emit a span instead).

> **GetPropagator(type) -> (inject, extract)**  

Author: @Oberon00 
Following up, now that Named Tracers have been added:
https://github.com/open-telemetry/opentelemetry-specification/pull/264

At first glance, I this approach turns out to make it a little tricky to implement an extensible inject/extract call which is used by multiple separate systems (observability, baggage, etc), as the observability extractor you want is now contextualized by the tracer handle. 

I can also see baggage wanting a similar scoping to tracers - surely users will want to disable certain endpoints for propagating baggage. I'm also noting that for any given code component, we would be naming tracers, meters, baggage, etc, essentially all with the same name and version.

So, while I like the named tracer concept and believe it is good, I'm wondering if this concept needs to be adjusted in some way? Otherwise, I'm not sure how we can cleanly split baggage and context propagation out from tracing.

To register with the propagation system, the Baggage API provides a set of propagation functions for every propagation type.

## Additional APIs

Because the application and context propagation layers are separated, it is possible to create new distributed applications which do not depend on either the Observability or Baggage APIs.

> **GetPropagator(type) -> (inject, extract)**  

Reviewer: same comment as above - the actual relationship is inverted, there's no actor that would ever call GetPropagator

To register with the propagation system, additional APIs provide a set of propagation functions for every propagation type.

# Context Propagation Layer

## Context API

Distributed applications access data in-process using a shared context object. Each distributed application sets a single key in the context, containing all of the data for that system.

**SetValue(context, key, value) -> context**  
To record the local state of an application, the Context API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

**GetValue(context, key) -> value**  
To access the local state of an application, the Context API provides a function which takes a context and a key as input, and returns a value.

### Optional: Automated Context Management

> When possible, context should automatically be associated with program execution . Note that some languages do not provide any facility for setting and getting a current context. In these cases, the user is responsible for managing the current context. 

Reviewer 2: might be worth expanded on what Automated Context Management is like in practice. I would imagine A ThreadLocal is an example since context is not available outside of the thread it was set or modified in?

Reviewer 3: 👍  In general, this OTEP is very abstract, some (non-normative) examples would be helpful.

Reviewer: This section is confusing two different layers from the [Context Propagation Layers design doc][1]: the Context and In-Process Propagation, hence the problem. In languages like Go the In-Process Propagation layer is not relevant, because the context is passed explicitly. This could apply to other frameworks when people don't want to rely on thread-locals. That has nothing to do with the Context type itself that simply stores values.

[1]: https://docs.google.com/document/d/1UxrEYOaQlF_E4gtiPoFmcZ4YKKe1GxohvCvQDuwvD1I/

Reviewer 3: I don't think the section is confusing them. It just defines a point where they interact. The context type used here is not necessarily the same the in-process layer uses, it could translate between them.

**SetCurrent(context)**  
To associate a context with program execution, the Context API provides a function which takes a Context.

**GetCurrent() -> context**  
To access the context associated with program execution, the Context API provides a function which takes no arguments and returns a Context.

## Propagation API

Distributed applications send data to downstream processes via propagators, functions which read and write application context into RPC requests. Each distributed application creates a set of propagators for every type of supported medium - currently HTTP and Binary.

**Inject(context, request)**  
To send the data for all distributed applications downstream to the next process, the Propagation API provides a function which takes a context and a request, and mutates the request to include the encoded context. The canonical representation of a request is as a map.

**Extract(context, request) -> context**  

> To receive data injected by prior upstream processes, the Propagation API provides a function which takes a context and a request, and returns an updated context.

Reviewer 4: What is purpose of the supplied `context`? Do we not just want to decode the context from the `request` to return a `Context`? If the context is also supplied - does it need to explicitly state how duplicate fields would be handled (i.e. existing context and request containing same key).

> To receive data injected by prior upstream processes, the Propagation API provides a function which takes a context and a request, and returns an updated context.

Reviewer 5: What does Extract update the context with? Is it SpanContext or Span?
If it is SpanContext then it would require another api to set the current SpanContext in addition to Span.
If it is Span then Extract also has to start a span which I don't think it is appropriate. 
So, it would be better for this API to return SpanContext. Unless we introduce a concept of Inactive Span which simply contains remote SpanContext.

Reviewer 3: You are right, I also think this is an important detail that needs to be resolved.

I have another suggestion: The context always contains just a spancontext. Any SpanContext has an optional reference to the span it refers to, which is null in case of a remote span (cf https://github.com/open-telemetry/opentelemetry-specification/pull/216#issuecomment-527379659)

Reviewer: I made a similar suggestion when we were working on the initial Java API in the spring. I think it has the benefits not just in the context of this discussion, but also as a alternative mechanism for implementing the "span". For example, in a streaming implementation the span does not even need to exist if all write operations simply work off the SpanContext. In other cases, spans may be kept by the tracer in a private dictionary, which makes memory management easier than if user code has an explicit reference to the span.

Reviewer 3: I don't understand how this would reduce references to Spans. I thought the user would be able to get this reference anytime from the SpanContext. But even withot such an API, startSpan etc would of course still return references to spans.

Reviewer: StartSpan would return a new span context. Span is just an accumulator of data, it is not actually needed in the API. We could move all Span methods into Tracer directly and make them accept span context as the first argument, that would be functionally equivalent.

Reviewer 3: Ah, I understand! There is one functional difference though: if you allowed something like `Tracer.setSpanAttribute(context, "x", "y")`, you implicitly allow setting attributes on spans that were not created in this process. This is prevented by the current Span/SpanContext separation and I think that's a good thing.

Reviewer 6: I remember that discussion on last spring, and I think the desire was that `Span`s created in the same process could share the monotonic clock (which is an optimization, more than a requirement).

Reviewer: @Oberon00 indeed, you can extract a parent span context and set attributes on it. I think it's actually a feature of this approach, not the downside :-) . Although it would only work in the streaming-like implementations, not the current approach where "finished span is not writeable". But then, even today you can do `span.end(); span.setTag()`, which is meaningless.

@carlosalberto there are many ways of implementing SDK when the API only exposes Span Context. As @Oberon00 said, the simple implementation is still to have the Span at the SDK level which is linked from Span Context, this way you can share clock offsets.

* **HTTP** - context is written into and read from a map of HTTP headers.
* **Binary** - context is serialized into and deserialized from a stream of bytes.

# Internal details

![drawing](img/context_propagation_details.png)

## Context details
OpenTelemetry currently implements three context types of context propagation.

**Span Context -** The serializable portion of a span, which is injected and extracted. The readable attributes are defined to match those found in the  W3C **traceparent** header. 

**Correlation Context -** Correlation Context contains a map of labels and values, to be shared between metrics and traces. This allows observability data to be indexed and dimensionalized in a variety of ways. Note that correlations can quickly add overhead when propagated in-band. But because this data is write-only, it may be possible to optimize how it is transmitted.

**Baggage Context -** Transaction-level application data, meant to be shared with downstream components. This data is readable, and must be propagated in-band. Because of this, Baggage should be used sparingly, to avoid ballooning the size of RPC requests.

> Note that when possible, OpenTelemetry APIs calls are given access to the entire context object, and not a specific context type.

Reviewer 2: I like the whole context object as it allows more flexibility for the propagators to modify contexts, but there's implementation details that all propagators will have to be aware of:

1. how to get and set specific contexts from the Context object. One way to accomplish this is with an explicit interface or methods attached to context, but as I understand it, this violates the idea that Yuri was evangelizing where the Context object is a primitive and is completely unaware of what values are being loaded into it.

2. SpanContext is a little tricky where the extractor should be setting the parent SpanContext, while the injector should be injecting values from the current SpanContext. This would be great to not require the propagator author to get that right. 

I'm personally voting for an explicit interface on context to retrieve such values. solving #2 is a little trickier. 

More context in this discussion:
https://github.com/open-telemetry/oteps/pull/37
And this PR:
https://github.com/open-telemetry/opentelemetry-python/pull/89

Reviewer: @toumorokoshi in Go you can define a private namespace for the values you store in Context ([example in go][1]). Nobody else but your cross-cutting application needs to know how to get/set those values. Therefore, the Context can remain completely agnostic to what it stores and carries.

[1]: https://github.com/opentracing/opentracing-go/blob/master/gocontext.go#L5

Reviewer 2: Ok, I think I understand what you're proposing. It's

1. a property-agnostic context object.
2. a layer of getter and setters to get and set values from that context. 

I'm fine with this approach. We actually had a discussion in the python group and that's the consensus we've reached as well.

Correlation values are solely to be used as labels for metrics and traces. By making Correlation data write-only, how and when it is transmitted remains undefined. This leaves the door open to optimizations, such as propagating some data out-of-band, and situations where sampling decisions may cease the need to propagate correlation context any further.

Baggage values, on the other hand, are explicitly added in order to be accessed by downstream by other application code. Therefore, Baggage Context must be readable, and reliably propagated in-band in order to accomplish this goal.

There may be cases where a key-value pair is propagated as TagMap for observability and as a Baggage for application specific use. AB testing is one example of such use case. There is potential duplication here at call site where a pair is created and also at propagation. 

Solving this issue is not worth having semantic confusion with dual purpose. However, because all observability functions take the complete context as input, it may still be possible to use baggage values as labels.

## What about complex propagation behavior?

> Some OpenTelemetry proposals have called for more complex propagation behavior. For example, falling back to extracting B3 headers if W3C Trace-Context headers are not found. Chained propagators and other complex behavior can be modeled as implementation details behind the Propagator interface. Therefore, the propagation system itself does not need to provide chained propagators or other additional facilities.

Reviewer 2: It sounds like a pretty universal need across languages to have multiple propagators. Any migration from one standard to another will require it. I'd personally rather see a standard need like this make it's way into the API rather than an unspecified SDK implementation whose behavior may differ across languages.

Reviewer 2: But I think this should just be a separate RFC regardless. This matches the current spec.

## Did you add a context parameter to every API call because Go has infected your brain?

No. The concept of an explicit context is fundamental to a model where independent distributed applications share the same context propagation layer. How this context appears or is expressed is language specific, but it must be present in some form.

> # Prior art and alternatives

Reviewer: ahem... Context Propagation Layers https://docs.google.com/document/d/1UxrEYOaQlF_E4gtiPoFmcZ4YKKe1GxohvCvQDuwvD1I/edit

# Proposal: Separate Layer for Context Propagation 

Status: `proposed`

Design OpenTelemetry as a set of separate applications which operate on a shared context propagation mechanism.

## Motivation

Based on prior art, we know that fusing the observability system and the context propagation system together creates issues. Observability systems have special rules for propagating information, such as sampling, and may have different requirements from other systems which require non-local information to be sent downstream.  
* Separation of concerns

>   *   Remove dependecy the Tracer dependency from context propagation mechanisms.

Reviewer 7: Delete `dependecy` before `the Tracer`

Author: whoop!

  *   For example: A/B testing, encrypted or authenticated data, and new, experimental forms of observability.

## Explanation

# OpenTelemetry Layered Architecture

![drawing](img/context_propagation_explanation.png)

OpenTelemetry is a distributed program, which requires non-local, transaction-level context in order to execute correctly. Transaction-level context can also be used to build other distributed programs, such as security, versioning, and network switching programs.

To allow for this extensibility, OpenTelemetry is separated into **application layer** and a **context propagation layer**.  In this architecture, multiple distributed applications - such as the observability and baggage systems provided by OpenTelemetry - simultaneously share the same underlying context propagation system in order to execute their programs.

# Application Layer

## Observability API

OpenTelemetry currently contains two observability systems - Tracing and Metrics – and may be extended over time. These separate systems are bound into a unified Observability API through sharing labels – a mechanism for correlating independent observations – and through sharing propagators.

> **Observe( context, labels…, observations...) context**

Author: Thanks @Oberon00. I'd like to experiment with the best approach for spec syntax. I like your suggestions, as it makes it more clear that this isn't supposed to be pseudo-go. :)

> **Observe( context, labels…, observations...) context**

Author: changed

The general form for all observability APIs is a function which takes a Context, label keys, and observations as input, and returns an updated Context.

**Correlate( context, label, value, hoplimit) context**
To set the label values used by all observations in the current transaction, the Observability API provides a function which takes a context, a label key, a value, and a hoplimit, and returns an updated context. If the hoplimit is set to NO_PROPAGATION, the label will only be available to observability functions in the same process. If the hoplimit is set to UNLIMITED_PROPAGATION, it will be available to all downstream services.

> **GetPropagator( type) inject, extract**

Author: GetPropagator returns an inject and extract function, with definitions matching those above. I'd rather write it like this than define objects and methods. But again, the syntax could be better.

perhaps:
```
GetPropagator( type) -> (inject, extract)
```

> **GetPropagator( type) inject, extract**

Author: done!

To register with the propagation system, the Observability API provides a set of propagation functions for every propagation type. 

## Baggage API

In addition to observability, OpenTelemetry provides a simple mechanism for propagating arbitrary data, called Baggage. This allows new distributed applications to be implemented without having to create new propagators.

To manage the state of a distributed application, the Baggage API provides a set of functions which read, write, and remove data.

**SetBaggage(context, key, value) context**
To record the distributed state of an application, the Baggage API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

> **GetBaggage( context, key) value**

Reviewer 8: ```suggestion
**GetBaggage(context, key) value**
```

To access the distributed state of an application, the Baggage API provides a function which takes a context and a key as input, and returns a value.

**RemoveBaggage( context, key) context**
To delete distributed state from an application, the Baggage API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

> **CleanBaggage( context) context**

Reviewer 6: I'd rather prefer `ClearBaggage` to keep the desired effect.

> **CleanBaggage( context) context**

Author: Thanks, `ClearBaggage` sounds better.

To avoid sending baggage to an untrusted downstream process, the Baggage API provides a function remove all baggage from a context.

> **GetPropagator( type) inject, extract**

Reviewer 8: ```suggestion
**GetPropagator(type) inject, extract**
```

To register with the propagation system, the Baggage API provides a set of propagation functions for every propagation type.

## Additional APIs

Because the application and context propagation layers are separated, it is possible to create new distributed applications which do not depend on either the Observability or Baggage APIs.

> **GetPropagator(type) inject, extract**

Author: So, the propagation layer provides the RegisterPropagator function. Applications provide, and applications provide the Propagators to be registered. 

Is there a better way to explain this in the GetPropagator description? Right now it says "To register with the propagation system, the [BLANK] API provides a set of propagation functions for every propagation type."

> **GetPropagator(type) inject, extract**

Author: I've removed the Registry concept, hopefully this is clearer.

To register with the propagation system, additional APIs provide a set of propagation functions for every propagation type.

# Context Propagation Layer

## Context API

Distributed applications access data in-process using a shared context object. Each distributed application sets a single key in the context, containing all of the data for that system.

**SetValue( context, key, value) context**  
To record the local state of an application, the Context API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.

**GetValue( context, key) value**  
To access the local state of an application, the Context API provides a function which takes a context and a key as input, and returns a value.

> **Optional: Automated Context Management**  

Author: I changed this to an actual subheading. Does that read better?

> **Optional: Automated Context Management**  

Reviewer 3: I think it does 👍

**Optional: SetCurrentContext( context)**  
To associate a context with program execution, the Context API provides a function which takes a Context.

**Optional: GetCurrentContext() context**  
To access the context associated with program execution, the Context API provides a function which takes no arguments and returns a Context.

## Propagation API

Distributed applications send data to downstream processes via propagators, functions which read and write application context into RPC requests. Each distributed application creates a set of propagators for every type of supported medium - currently HTTP and Binary.

**Inject( context, request)**  
To send the data for all distributed applications downstream to the next process, the Propagation API provides a function which takes a context and a request, and mutates the request to include the encoded context. The canonical representation of a request is as a map.

**Extract( context, request) context**  

> To receive the data from all distributed applications set by prior upstream processes, the Propagation API provides a function which takes a context and a request, and mutates the request to include the encoded context. The canonical representation of a request is as a map.

Reviewer 7: Everything after the first comma is identical to the corresponding `Inject` text. I think you mean it `mutates the context`?

Author: Fixed. In general, the descriptions of the functions could be better, I'll improve them this week. Please let me know if the functions themselves capture everything.

OpenTelemetry currently contains two types of Propagators:

* **HTTP** - context is written into and read from a map of HTTP headers.
* **Binary** - context is serialized into and deserialized from a stream of bytes.

# Internal details

![drawing](img/context_propagation_details.png)

## Context details
OpenTelemetry currently implements three context types of context propagation.

**Span Context -** Observability data used by the tracing system. The readable attributes are defined to match those found in the  W3C **traceparent** header. Span Context is used as labels for metrics and traces. This can quickly add overhead when propagated in-band. But, because this data is write-only, how this information is transmitted remains undefined. 

> **Correlation Context -** Transaction-level observability data, which can be applied as labels to spans and metrics. 

Reviewer 9: can we better explain the difference between correlation context and span context?

Author: Yeah totally. Please let me know if the new descriptions are better.

Reviewer 9: Much better.

**Baggage Context -** Transaction-level application data, meant to be shared with downstream components.

Note that when possible, OpenTelemetry APIs calls are given access to the entire context object, and not a specific context type.

## Context Management and in-process propagation

> In order for Context to function, it must always remain bound to the execution of code it represents. By default, this means that the programmer must pass a Context down the call stack as a function parameter.  However, many languages provide automated context management facilities, such as thread locals.  OpenTelemetry should leverage these facilities when available, in order to provide automatic context management.

Reviewer 8: Minor, there are mixed use of single space and double spaces.

Author: Thank you for all the copyediting!

## Pre-existing Context implementations

In some languages, a single, widely used Context implementation exists. In other languages, there many be too many  implementations, or none at all. In the cases where there is not an extremely clear pre-existing option available, OpenTelemetry should provide its own Context implementation.

While the above explanation represents the default OpenTelemetry approach to context propagation, it is important to note that some languages may already contain a form of context propagation. For example, Go has a the context.Context object, and widespread conventions for how to pass.

> Span data is used as labels for metrics and traces. This can quickly add overhead when propagated in-band. But, because this data is write-only, how this information is transmitted remains undefined. 

Author: derp, this is just a crappy unfinished section.

> Span data is used as labels for metrics and traces. This can quickly add overhead when propagated in-band. But, because this data is write-only, how this information is transmitted remains undefined. 

Author: Agree, it's filled in better below, under "Why separate Baggage from Correlations?" I'll delet it from here.

> Span data is used as labels for metrics and traces. This can quickly add overhead when propagated in-band. But, because this data is write-only, how this information is transmitted remains undefined. 

Author: DELETED!

Correlations values are solely to be used as labels for metrics and traces. By making Correlation Context data write-only, how and when it is transmitted remains undefined. This leaves the door open to optimizations, such as propagating some data out-of-band, and situations where sampling decisions may cease the need to propagate correlation context any further.

Baggage values, on the other hand, are explicitly added in order to be accessed by downstream by other application code. Therefore, Baggage Context must be readable, and reliably propagated in-band in order to accomplish this goal.

There may be cases where a key-value pair is propagated as TagMap for observability and as a Baggage for application specific use. AB testing is one example of such use case. There is potential duplication here at call site where a pair is created and also at propagation. 

Solving this issue is not worth having semantic confusion with dual purpose. However, because all observability functions take the complete context as input, it may still be possible to use baggage values as labels.

## What about complex propagation behavior?

> Some OpenTelemetry proposals have called for more complex propagation behavior. For example, having a fallback to extracting B3 headersif Trace-Context headers are not found. Chained propagators and other complex behavior can be modeled as implementation details behind the Propagator interface. Therefore, the propagation system itself does not need to provide chained propagators or other additional facilities.

Reviewer 3: ```suggestion
Some OpenTelemetry proposals have called for more complex propagation behavior. For example, falling back to extracting B3 headers if W3C Trace-Context headers are not found. Chained propagators and other complex behavior can be modeled as implementation details behind the Propagator interface. Therefore, the propagation system itself does not need to provide chained propagators or other additional facilities.
```

> Some OpenTelemetry proposals have called for more complex propagation behavior. For example, having a fallback to extracting B3 headersif Trace-Context headers are not found. Chained propagators and other complex behavior can be modeled as implementation details behind the Propagator interface. Therefore, the propagation system itself does not need to provide chained propagators or other additional facilities.

Reviewer 3: > the propagation system itself does not need to provide chained propagators

So what happens when multiple propagators are registered for the same type? I think chained propagators may be cumbersome to provide on top of an API that allows only one propagator, since it requires cooperation between everyone who wants to add a propagator to the chain.

Reviewer 6: I have seen code that does this, by using a stack-alike `Propagator` that tries to inject/extract from a list of propagators (order is important) - for extraction, it returns with the first successful attempt, and for injection it simply tries to inject all formats. And yes, we should have a test case for this scenario, so we know where work fine.

Author: So, here is the nuance: there is basic chaining provided at the API level by RegisterPropagator. All propagators are run for every type, in the order in which they are registered. This is where the cooperation happens between independent applications.

More complex propagation behavior usually ends up being specific to each application. So, the OTel SDK can provide the kind of fallback W3C -> B3 chaining for observability, described here. That sort of behavior is presented as a single, complex propagator, from the point of view of the Propagation API, since the fallback behavior is internal to a single application. Does that make sense?

> ## Did you add a context parameter to every API call because Go has infected your brain?

Reviewer 8: Curious, is this just for fun?

Author: Haha, well I wanted to address what I perceive to be a common question, along the lines of "why is this context parameter everywhere? Is it because this is a golang project? Is it required that every language must pass context this way?"

But if the humor gets in the way of learning, I can change it.

Reviewer: I like it. You can also stress that it's not just a Go thing - in the old versions of Node there was no CLS (or it was extremely inefficient) and explicitly passing the context was how the propagation was achieved.

Author: Good call, I will emphasize that this issue exists in multiple languages.

No. The concept of an explicit context is fundamental to a model where independent distributed applications share the same context propagation layer. How this context appears or is expressed is language specific, but it must be present in some form.

# Prior art and alternatives

Prior art:  
* OpenTelemetry distributed context
* OpenCensus propagators
* OpenTracing spans
* gRPC context

# Open questions

Related work on HTTP propagators has not been completed yet.

* [W3C Trace-Context](https://www.w3.org/TR/trace-context/) candidate is not yet accepted
* Work on [W3C Correlation-Context](https://w3c.github.io/correlation-context/) has begun, but was halted to focus on Trace-Context. 
* No work has begun on a theoretical W3C Baggage-Context.

Given that we must ship with working propagators, and the W3C specifications are not yet complete, how should we move forwards with implementing context propagation?

# Future possibilities

> Cleanly splitting OpenTelemetry into an Application and  Context Propagation layer may allow us to move the Context Propagation layer into its own, stand-alone project. This may facilitate adoption, by allowing us to share Context Propagation with gRPC and other projects.

Reviewer 8: ```suggestion
Cleanly splitting OpenTelemetry into an Application and Context Propagation layer may allow us to move the Context Propagation layer into its own, stand-alone project. This may facilitate adoption, by allowing us to share Context Propagation with gRPC and other projects.
```
